Firestore/Source/Remote/FSTExponentialBackoff.mm

```diff
@@ -69,13 +72,32 @@ - (void)backoffAndRunBlock:(void (^)(void))block {
   [self cancel];
 
   // First schedule the block using the current base (which may be 0 and should be honored as such).
-  NSTimeInterval delayWithJitter = _currentBase + [self jitterDelay];
+  NSTimeInterval desiredDelayWithJitter = _currentBase + [self jitterDelay];
+
+  // Guard against lastAttemptTime being in the future due to a clock change.
+  NSTimeInterval delaySoFar = MAX(0, [[NSDate date] timeIntervalSince1970] - self.lastAttemptTime);
+
+  // Guard against the backoff delay already being past.
+  NSTimeInterval remainingDelay = MAX(0, desiredDelayWithJitter - delaySoFar);
+
   if (_currentBase > 0) {
-    LOG_DEBUG("Backing off for %s seconds (base delay: %s seconds)", delayWithJitter, _currentBase);
+    LOG_DEBUG(
+        "Backing off for %s seconds ("
+        "base delay: %s seconds, "
+        "delay with jitter: %s seconds, "
+        "last attempt: %s seconds ago)",
+        remainingDelay, _currentBase, desiredDelayWithJitter, delaySoFar);
   }
 
-  self.timerCallback =
-      [self.dispatchQueue dispatchAfterDelay:delayWithJitter timerID:self.timerID block:block];
+  FSTWeakify(self);
+  self.timerCallback = [self.dispatchQueue dispatchAfterDelay:remainingDelay
+                                                      timerID:self.timerID
+                                                        block:^{
+                                                          FSTStrongify(self);
+                                                          self.lastAttemptTime =
```

Comment:
If `self` has become `nil` the the assignment will nil-dispatch so strictly speaking that's okay, but should we really call the block?

I think this should be:

```objc
^{
  FSTStrongify(self);
  if (self) {
    self.lastAttemptTime = [[NSDate date] timeIntervalSince1970];
    block();
  }
}
```

Reply:
Sounds reasonable. Fixed.